tests/run.sh

```diff
@@ -70,34 +80,50 @@ python <<-EOF >"${CLUSTER_NAME}.yaml"
 
 	import yaml
 
-	with open('examples/tectonic.aws.yaml') as f:
+	with open('examples/tectonic.${BACKEND}.yaml') as f:
 	    config = yaml.load(f)
 	config['name'] = '${CLUSTER_NAME}'
 	with open(os.path.expanduser(os.path.join('~', '.ssh', 'id_rsa.pub'))) as f:
 	    config['admin']['sshKey'] = f.read()
 	config['baseDomain'] = '${DOMAIN}'
 	with open('${PULL_SECRET_PATH}') as f:
 	    config['pullSecret'] = f.read()
-	config['aws']['region'] = '${AWS_REGION}'
-	config['aws']['extraTags'] = {
-	    'expirationDate': (
-	        datetime.datetime.utcnow() + datetime.timedelta(hours=4)
-	    ).strftime('%Y-%m-%dT%H:%M+0000'),
-	}
-	if ${CONFIGURE_AWS_ROLES:-False}:
-	    config['aws']['master']['iamRoleName'] = 'tf-tectonic-master-node'
-	    config['aws']['worker']['iamRoleName'] = 'tf-tectonic-worker-node'
+	if '${BACKEND}' == 'aws':
+	    config['aws']['region'] = '${AWS_REGION}'
+	    config['aws']['extraTags'] = {
+	        'expirationDate': (
+	            datetime.datetime.utcnow() + datetime.timedelta(hours=4)
+	        ).strftime('%Y-%m-%dT%H:%M+0000'),
+	    }
+	    if ${CONFIGURE_AWS_ROLES:-False}:
+	        config['aws']['master']['iamRoleName'] = 'tf-tectonic-master-node'
+	        config['aws']['worker']['iamRoleName'] = 'tf-tectonic-worker-node'
+	elif '${BACKEND}' == 'libvirt':
+	    config['libvirt']['imagePath'] = '${IMAGE_PATH}'
 	yaml.safe_dump(config, sys.stdout)
 	EOF
 
 echo -e "\\e[36m Initializing Tectonic...\\e[0m"
 tectonic init --config="${CLUSTER_NAME}".yaml
 
+trap destroy EXIT
```

Comment:
While we're at it, how about trapping INT (control+c) too?

Reply:
> While we're at it, how about trapping INT (control+c) too?

I think `EXIT` covers that:

```console
$ cat /tmp/test.sh
#!/bin/sh

testing()
{
  echo testing
}

trap testing EXIT
sleep 10
$ /tmp/test.sh
^Ctesting
```

Reply:
I haven't found out why but this doesn't work in our case, i.e. the cluster isn't destroyed if you cancel the tests using ^C.

Reply:
> I haven't found out why but this doesn't work in our case, i.e. the cluster isn't destroyed if you cancel the tests using ^C.

Looks like it works to me:

```console
$ ./tests/run.sh libvirt
...
 Creating Tectonic configuration...
 Initializing Tectonic...
 Deploying Tectonic...
^C Exiting... Destroying Tectonic...
 Finished! Smoke test output: Never executed. Problem with one of previous stages
 So Long, and Thanks for All the Fish
```

You can see the output from [here][1] in that session.  If you're seeing leaked resources, my guess is that you have a SIGINT in the middle of our multi-step Terraform initialization, and our multi-step Terraform destruction code is choking and dying on the partially initialized cluster.

[1]: https://github.com/openshift/installer/blob/d316449bbf420926b4d927ec7e29d19828e531f3/tests/run.sh#L20

Reply:
I don't see the those echo messages printed when I cancel and I have tried multiple times. I can't reproduce this in a MWE though, will have to recheck again tomorrow.

Reply:
> I don't see the those echo messages printed when I cancel...

Do you SIGINT during the smoke tests themselves (after setup completes) with `LEAVE_RUNNING=y`?  Then you'd be sending it after [we cleared the handler][1].  If you're not seeing the cleanup code in some other place, can you post your `*.log` (from [this `tee`][2]) and possibly an `strace` of the script?

[1]: https://github.com/openshift/installer/pull/231/commits/26d3a92e9d8b076ccaa31b9f213d7fa687c1fe19#diff-76f9497ef7c0797cbd5e943cf295c9d1R128
[2]: https://github.com/openshift/installer/blob/26d3a92e9d8b076ccaa31b9f213d7fa687c1fe19/tests/run.sh#L19

Reply:
I tested cancelling the script while the smoke test were run and explicitly set `LEAVE_RUNNING=n`.

```
libvirt_domain.bootstrap: Creation complete after 2s (ID: 59e09c3a-f972-474b-b2b8-d49b636d1397)

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
 Running smoke test...
=== RUN   Test
=== RUN   Test/Common
=== RUN   Test/Common/APIAvailable
^C
steveej@steveej-laptop ✗ (130) ~/src/go/src/github.com/openshift/installer  4019d63a ⚡
```

Reply:
> I tested cancelling the script while the smoke test were run and explicitly set `LEAVE_RUNNING=n`.

It works for me:

```console
Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
 Running smoke test...
=== RUN   Test
=== RUN   Test/Common
=== RUN   Test/Common/APIAvailable
^C Exiting... Destroying Tectonic...
data.terraform_remote_state.assets: Refreshing state...
```

That diff does not look related, because the trap is in the Bash process.  To test that hypothesis, try replacing `./smoke -test.v --cluster` with `(echo sleeping; sleep 10; echo slept)`

I don't think this is going to be a Bash issue, but just in case, what version are you running?

```console
$ /usr/bin/env bash --version
GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)
...
```

Can you keep trimming parts of `run.sh` out to make it look more like [my skeleton][1] and see if/when it starts working for you?

[1]: https://github.com/openshift/installer/pull/231#discussion_r216701223

Reply:
Here's a [MWE](https://gist.github.com/steveeJ/86efe22e8d2195f5d19efe05d03225b2#file-run-sh) and an strace that points out that any [write to stdout leads to EPIPE within the destroy function](https://gist.github.com/steveeJ/86efe22e8d2195f5d19efe05d03225b2#file-strace-log-L2359):

> 15957 write(1<pipe:[8516319]>, "m4g1k\n", 6) = -1 EPIPE (Broken pipe)

Reply:
> Here's a MWE and...

Lovely, thanks :).  Looking at the strace:

```console
$ grep 'killed\|EPIPE\|8516319' strace.log  | tail -n13
15959 read(0<pipe:[8516319]>,  <unfinished ...>
15959 +++ killed by SIGINT +++
15961 +++ killed by SIGINT +++
15964 +++ killed by SIGINT +++
15962 +++ killed by SIGINT +++
15963 +++ killed by SIGINT +++
15960 +++ killed by SIGINT +++
15958 +++ killed by SIGINT +++
15957 write(1<pipe:[8516319]>, "m4g1k\n", 6) = -1 EPIPE (Broken pipe)
15957 rt_sigreturn({mask=[]})           = -1 EPIPE (Broken pipe)
15957 write(2<pipe:[8516319]>, "./tests/run.sh: line 7: echo: wr"..., 55) = -1 EPIPE (Broken pipe)
15957 rt_sigreturn({mask=[]})           = -1 EPIPE (Broken pipe)
15957 +++ killed by SIGINT +++
```

I'm decoding those PIDs now so I understand how that SIGINT got passed around.  Or did you ctrl-C multiple times?

Reply:
AFAIR I only pressed once

Reply:
I certainly don't know, I'm not sure about you ;-) I'd regard this as something that should just work.

Reply:
Fix confirmed! Awesome :tada: